File: java-to-go-projects/04-log-parser/rust/src/main.rs
use regex::Regex;
use std::collections::{HashMap, HashSet};
use std::env;
use std::fs;

struct LogEntry {
    timestamp: String,
    level: String,
    path: String,
    message: String,
}

fn parse_file(filename: &str) -> Vec<LogEntry> {
    let content = fs::read_to_string(filename).expect("Failed to read log file");
    let re = Regex::new(r"\[(.*?)\] (\w+)\s+(\S+) - (.*)").unwrap();

    content
        .lines()
        .filter_map(|line| {
            let caps = re.captures(line)?;
            Some(LogEntry {
                timestamp: caps[1].to_string(),
                level: caps[2].to_string(),
                path: caps[3].to_string(),
                message: caps[4].to_string(),
            })
        })
        .collect()

}



fn count_by_level(entries: &[LogEntry]) -> HashMap<String, usize> {
    let mut counts = HashMap::new();
    for entry in entries {
        *counts.entry(entry.level.clone()).or_insert(0) += 1;
    }
    counts
}

fn filter_by_level<'a>(entries: &'a [LogEntry], level: &str) -> Vec<&'a LogEntry> {
    entries.iter()
        .filter(|e| e.level.eq_ignore_ascii_case(level))
        .collect()
}

fn unique_paths(entries: &[LogEntry]) -> HashSet<String> {
    entries.iter().map(|e| e.path.clone()).collect()
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: {} <logfile>", args[0]);
        return;
    }

    let entries = parse_file(&args[1]);
    println!("Total log entries: {}", entries.len());

    let counts = count_by_level(&entries);
    for (level, count) in counts {
        println!("{}: {}", level, count);
    }

    let error_entries = filter_by_level(&entries, "ERROR");
    println!("Error entries:");
    for entry in error_entries {
        println!("[{}] {} - {}", entry.timestamp, entry.path, entry.message);
    }

    let paths = unique_paths(&entries);
    println!("Unique paths:");
    for path in paths {
        println!("{}", path);
    }
}

File: java-to-go-projects/04-log-parser/go/main.go
package logparser
package main

import (
	"fmt"
	"os"
	"regexp"
)

type LogEntry struct {
	Timestamp string
	Level     string
	Path      string
	Message   string
}

// parseFile reads a log file and returns parsed entries.
func parseFile(filename string) []LogEntry {
	panic("todo")
}

// countByLevel returns a map of log level -> count.
func countByLevel(entries []LogEntry) map[string]int {
	panic("todo")
}

// filterByLevel returns entries matching the given level (case-insensitive).
func filterByLevel(entries []LogEntry, level string) []LogEntry {
	panic("todo")
}

// uniquePaths returns the set of unique request paths.
func uniquePaths(entries []LogEntry) []string {
	panic("todo")
}

func main() {
	if len(os.Args) < 2 {
		fmt.Fprintf(os.Stderr, "Usage: %s <logfile>\n", os.Args[0])
		os.Exit(1)
	}

	_ = regexp.MustCompile(`\[(.*?)\] (\w+)\s+(\S+) - (.*)`)

	panic("todo: parse file, print counts, errors, unique paths")
}
